library/src/main/java/com/bumptech/glide/load/ImageHeaderParserUtils.java

```diff
@@ -31,9 +31,7 @@ public static ImageType getType(List<ImageHeaderParser> parsers, @Nullable Input
     }
 
     is.mark(MARK_POSITION);
-    int size = parsers.size();
-    for (int i = 0; i < size; i++) {
-      ImageHeaderParser parser = parsers.get(i);
+    for (ImageHeaderParser parser : parsers) {
```

Comment:
These are deliberate. Using for each allocates an iterator, which we avoid by writing the loop ourselves. Same with the rest of the loops unfortunately.

Reply:
Awwww, really? Is that perf improvement necessary for _every_ other for loop or just the ones in this class? Also, do you mind me adding a comment explaining the perf improvement to each loop so other people don't make the same mistake?

Reply:
I usually default to starting with for each, then switch when I see the allocations in the allocation tracker during scrolling. 

On ART devices this kind of optimization is probably irrelevant, but avoiding object allocations can make a pretty big difference on dalvik devices. 

You're more than welcome to add a comment.

Reply:
Oh, got it. I'm happy as long as you start with the most readable version first. 😄

Reply:
Last question: is there any reason to cache `size` and not just call it repeatedly in the if statement as is the standard? It's just a field in `ArrayList` so calling it should be dirt cheap.

Reply:
It keeps the performance the same in both cases, at least theoretically. I don't have much evidence that it matters, and in most cases in Glide it's probably not called often enough to matter at all, but it's easy to write and not particularly less readable.

Effective java mentions that for each might be slightly more performant than a standard loop because it memoizes the size: https://stackoverflow.com/a/256861/800716.

We've also seen some interesting behavior in ART in particular where keeping local copies of variables around can make a huge difference in some versions of Android in some loops: e7a494253187f14bc73a909f522e4da2ce093920

Reply:
@sjudd Alrighty, thanks for answering all my questions—that's some really cool stuff! 😄